website/dynamo.py

```diff
@@ -298,42 +348,163 @@ def get_many(self, key, reverse=False, limit=None, pagination_token=None, filter
+        predicate = make_predicate(client_side_filter)
+
+        batch_size = math.ceil(limit * max(1.0, fetch_factor))
+
+        # We need to know if we're doing a prevpage query or not.
+        inverse_page, initial_pagination_token = decode_page_token(pagination_token)
+
+        curr_pagination_token = pagination_token
+        dropped_remaining_in_this_page = False
+        first_page = None
+        while len(items) < limit:
+            space_remaining = limit - len(items)
+
+            page = self.get_many(key, reverse=reverse, limit=batch_size,
+                                 pagination_token=curr_pagination_token, filter=server_side_filter)
+            if not first_page:
+                first_page = page
+            selected_in_this_page = [row for row in page if predicate(row)]
+
+            if inverse_page:
+                # They're already in the right order, but they need to go in the right place as well
+                items = selected_in_this_page[-space_remaining:] + items
+            else:  # Forward
+                items.extend(selected_in_this_page[:space_remaining])
+            dropped_remaining_in_this_page = len(selected_in_this_page) > space_remaining
+
+            curr_pagination_token = page.next_page_token if not inverse_page else page.prev_page_token
+            if not curr_pagination_token or cancel.is_cancelled():
+                break
+
+        if inverse_page:
+            if dropped_remaining_in_this_page:
+                prev_page_token = page.pagination_key.extract_dict(items[0])
+            else:
+                # Need to decode because it will be re-encoded below
+                prev_page_token = decode_page_token(page.prev_page_token)[1] if page.prev_page_token else None
+
+            # The last element of the first page marks our next_page_token. If this page is empty for
+            # some reason, we just pretend that the initial pagination token is our forward pagination token.
+            # Not entirely correct (we'll drop 1 element) but at least it gets us back into the flow.
+            next_page_token = page.pagination_key.extract_dict(
+                first_page[-1]) if first_page else initial_pagination_token
+        else:
+            if dropped_remaining_in_this_page:
+                next_page_token = page.pagination_key.extract_dict(items[-1])
+            else:
+                # Need to decode because it will be re-encoded below
+                next_page_token = decode_page_token(page.next_page_token)[1] if page.next_page_token else None
+
+            # The first element of the first page marks our prev_page_token
+            prev_page_token = page.pagination_key.extract_dict(
+                first_page[0]) if first_page and pagination_token else None
+
+        return ResultPage(
+            items,
+            encode_page_token(next_page_token, False),
+            encode_page_token(prev_page_token, True))
```

Comment:
would be nicer to switch the parameter's order!

Reply:
Yeah, might be. I'm always a bit hesitant with obviously backwards incompatible changes, but I guess in this case it might be okay...